---
title: "AdvancedRNA-seq"
author: "Simon Tomlinson"
date: "12/03/2021"
output:
  html_document: default
  pdf_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning=FALSE, message=FALSE)
knitr::opts_knit$set(root.dir = "~/FGT/W10_RNAseq/")
```

## Advanced RNA-Seq Data Analysis Using R
### Simon Tomlinson FGT 2021

-Create a directory to store this week’s files and unpack the archive file for tutorial8 into this folder. 
   tar -xvf FGT_T8_RNAseq_tutorial.tar
-Set this as the current working directory
-Set the root dir in the markdown (r setup chunk)
--eg as an example knitr::opts_knit$set(root.dir = "./fgt2022/tutorial8/FGT_T8_RNAseq_tutorial/")
-Then load and run this script

```{r init, echo=FALSE}
#Load the required libraries & load the files for tutorial8
#library(Rsubread)
library(edgeR)
library(limma)
library(gplots)
library(DESeq2)
library(affy)
library(QuasR)
load("FGT_T8_RNA-seq.Rdata")

#The input files were generated using this code- 
#alignment_files <-list.files(path = "./aligned_merged",pattern="*.bam",full.names=T)
#Generate a simple QC file to confirm the QC results

#cl <-makeCluster(15)
#qQCReport(alignment_files, pdfFilename="results/qc_report.pdf",clObj=cl,useSampleNames=T)
#stopCluster(cl)

#use featureCounts to load the BAM files into the 
#mytable_feaures <-featureCounts(files=alignment_files,annot.ext="./omes/annot/Mus_musculus.GRCm38.99.gtf",isGTFAnnotationFile = TRUE, countMultiMappingReads=F, minMQS=30, nthreads=10,minOverlap=10)


#Note- we generate “mytable_feaures” R list object which has slots (use “$” to access).  
#We also generate “qc_report.pdf” which contains extra QC information.

```
Once we have loaded the data we are ready to run some further analysis.

```{r input_processing, echo=FALSE}
#Some further processing to access count data and column names.
table_rnaseq <- (mytable_feaures)$counts
colnames_rnaseq <- colnames(table_rnaseq)
#tidy up the sample names eg"X.home.guest101.new_rnaseq.aligned.SRR5054353.sorted.bam"
#to ShortName  using the “ targets file called adf here
adf <- read.table("mychromosome_samples4.txt",sep='\t',
                  row.names=1,fill=T,header=T)
#This orders the elements in the table
idx <- match(colnames_rnaseq,rownames(adf))
colnames(table_rnaseq) <- adf$ShortName[idx]
adf <- adf[idx,]
rownames(adf) <- colnames(table_rnaseq)
```


Further processing to remove unwanted samples and create the required object for the differential gene analysis using DESeq (https://bioconductor.org/packages/release/bioc/html/DESeq.html).

```{r remove_unwanted_sampled, echo=TRUE}
#keep-remove XO samples
keep_cols=adf$Sex[idx]!="O"
#create object for DESeq
#NOTE the design for differential analysis is fed into this object- the adf annotates each sample 
dds_rnaseq <- DESeqDataSetFromMatrix(countData=table_rnaseq[,keep_cols],
                                     colData=adf[keep_cols,],
                                     design=~ Line + Sex)
#check the object dimensions
dim(dds_rnaseq)
head(rownames(dds_rnaseq))
colnames(dds_rnaseq)
#check the object type
class(dds_rnaseq)
typeof(dds_rnaseq)


#check what this object contains
slotNames(dds_rnaseq)
colnames(colData(dds_rnaseq))
colnames(assay(dds_rnaseq))

```


Apply filters to remove genes that do not pass quality filters.

```{r apply_filters, echo=FALSE}
#first filter based upon counts
keep <- rowSums(counts(dds_rnaseq)) > 1
dds_rnaseq <- dds_rnaseq[keep,]
nrow(dds_rnaseq)
#second filter-require >10 in >= 3 samples
keep <- rowSums(counts(dds_rnaseq) >= 10) >= 3
dds_rnaseq <- dds_rnaseq[keep,]
nrow(dds_rnaseq)

#size factor estimation for normalisation- these are used to scale by library size when data is exported.
dds_rnaseq <- estimateSizeFactors(dds_rnaseq)
sizeFactors(dds_rnaseq)
```

DESeq2 also provides methods that outputs data with various normalisations….  


```{r normalisation, echo=TRUE}
# Normalised log counts
counts_rnaseq <- log2(counts(dds_rnaseq, normalized=TRUE))
# fpm is similar to cpm (https://rdrr.io/bioc/DESeq2/man/fpm.html)
# Here the data is log transformed
fpm_rnaseq <- log(fpm(dds_rnaseq))
# this is also normalised by the sizeFactors
# rlog and vsd are discussed here https://genomebiology.biomedcentral.com/articles/10.1186/s13059-014-0550-8
vsd_rnaseq <- vst(dds_rnaseq, blind = T)
head(vsd_rnaseq, 3)
rld_rnaseq <- rlog(dds_rnaseq, blind = T)
head(rld_rnaseq, 3)
```

We will use rlog for the rest of this tutorial.  Note that differential expression uses counts- rlog is just for visualisation.

```{r plot_normalised, echo=FALSE}
#make some plots with normalised data
boxplot(counts_rnaseq,
        main="Counts")
boxplot(fpm_rnaseq,
        main="Fragments per million")
boxplot(assay(vsd_rnaseq),
        main="Variance-stabilizing transformed")
boxplot(assay(rld_rnaseq),
        main="Regularized-logarithm transformed")
#but suppose we look at non-normalised
na.rm=T
counts_un <-counts(dds_rnaseq, normalized=FALSE)
boxplot(log2(counts_un),
        main="Counts, non-normalized")
counts_un_rlog<-rlog(counts_un)

#make mva.pairs-this is slow “mva.pairs.png” is provided 
#png(filename="mva.pairs.png",width=2000, height=2000)
#mva.pairs(counts_un_rlog)
#dev.off()
```
*see mva.pairs.png for MA-plots of all samples.

PCA plots similar to those we use for microarrays.

```{r build_pca, echo=FALSE}
#we can run a PCA plot here...
library(scatterplot3d)
library(ggplot2)
# Perform PCA
pca <- prcomp(t(na.omit(assay(rld_rnaseq))), scale=T)
# Plot the PCA results
s3d<-scatterplot3d(pca$x[,1:3], 
                   pch=19, 
                   color=colData(dds_rnaseq)$Colour)
s3d.coords <- s3d$xyz.convert(pca$x[,1:3])
text(s3d.coords$x, s3d.coords$y, 
     labels = colnames(rld_rnaseq),
     pos = 3, offset = 0.5, cex=0.5)
#or a quick plot in 2d
qplot(pca$x[,1],pca$x[,2],xlab="PCA1",
      ylab="PCA2",color=colData(dds_rnaseq)$Colour)
#We can also output a distance matrix…
sampleDists <- dist(t(assay(rld_rnaseq)))
sampleDists
```

Generate further heatmaps with the data

```{r more_heatmaps, echo=FALSE}
library("pheatmap")
library("RColorBrewer")
sampleDistMatrix <- as.matrix(sampleDists)
pheatmap(sampleDistMatrix,
         main="Heatmap of samples distance matrix")
#easier PCA plot
plotPCA(rld_rnaseq, 
        intgroup = c("Sex", "Line"))
```

Differential Gene Expression\
Here we perform a two factor analysis using DESeq2.\
The design was fed into the dds_rnaseq object and was specified out as ~Line + Sex which specifies two factors Sex and Line.\
Sex and Line are colnames in the annotation able and samples are therefore annotated with these labels.  Statistics in DESeq2 works in a similar way to Limma but is designed for count data. 

```{r differential_expression, echo=FALSE}
dds_rnaseq <- DESeq(dds_rnaseq)
#This is how you specify a contrast…
result_sex <- results(dds_rnaseq,
                     contrast=c("Sex","M","F"))
result_cell_line <- results(dds_rnaseq,
                           contrast=c("Line","BC","CB"))
#make a nice plot of differential expression
plotMA(result_sex, main="DESeq2", ylim=c(-2,2))
table(result_sex$padj < 0.01)
table(result_cell_line$padj < 0.01)
table(result_sex$padj < 0.05)
table(result_cell_line$padj < 0.05)
result_sex_selected <- subset(result_sex, padj < 0.05)
result_sex_selected <- result_sex_selected[order(abs(result_sex_selected$log2FoldChange), decreasing=TRUE), ]
head(result_sex_selected)
```

Make heatmaps from each of the output tables- selected top 50 or all using fold change and FDR.

```{r differential_expression_heatmaps, echo=FALSE}
top50 <- rownames(result_sex_selected)[1:50]
#make a heatmap for the top 50
pheatmap(assay(rld_rnaseq)[top50,],
         scale="row",
         show_rownames=T,
         main="Heatmap of Top 50 DEGs: M versus F")
#all
pheatmap(assay(rld_rnaseq)[rownames(result_sex_selected),],
         scale="row",
         show_rownames=F,
         main="Heatmap of all DEGs: M versus F")

#and the cell lines
result_cell_selected <- subset(result_cell_line, padj < 0.05)
result_cell_selected <- result_cell_selected[order(abs(result_cell_selected$log2FoldChange), decreasing=TRUE), ]
head(result_cell_selected)
top50c <- rownames(result_cell_selected)[1:50]
#make a heatmap for the top 50
pheatmap(assay(rld_rnaseq)[top50c,],
         scale="row",
         show_rownames=T,
         main="Heatmap of Top 50 DEGs: Line BC vs CB")
#all
pheatmap(assay(rld_rnaseq)[rownames(result_cell_selected),],
         scale="row",
         show_rownames=F,
         main="Heatmap of all DEGs: Line BC vs CB")
```


Build Gene Annotation.  Ensembl may be down at the moment.  In this case the 'resultAnnot' object is provided

```{r build_gene_annotation, echo=TRUE}
#Download Ensembl annotation using BiomaRt and rename the samples
library(biomaRt)
#UK ensembl is being updated so we use a USA mirror, "useast.ensembl.org"
ensembl_host <- "uswest.ensembl.org"
head(biomaRt::listMarts(host = ensembl_host), 15)
head(biomaRt::listAttributes(biomaRt::useDataset(dataset = "mmusculus_gene_ensembl",mart = useMart("ENSEMBL_MART_ENSEMBL",host = ensembl_host))), 40) 
mart <- biomaRt::useDataset(dataset = "mmusculus_gene_ensembl",
                            mart = useMart("ENSEMBL_MART_ENSEMBL",
                                           host = ensembl_host))
#resultAnnot <- biomaRt::getBM(values=rownames(dds_rnaseq),attributes = c("ensembl_gene_id","external_gene_name","chromosome_name","start_position","end_position","description","strand"),filters="ensembl_gene_id",mart=mart)
```

Merge Annotation with Input Data

```{r merge_gene_annotation, echo=FALSE}
#merge with input data
names <- resultAnnot[,1]
resultAnnot <- as.data.frame(resultAnnot)
rownames(resultAnnot) = names
idx <- match(rownames(dds_rnaseq),rownames(resultAnnot))
#make sure annotation is in same order
all(rownames(dds_rnaseq) == rownames(resultAnnot))
grr <- resultAnnot[match(rownames(dds_rnaseq), resultAnnot$ensembl_gene_id),]
all(rownames(dds_rnaseq) == rownames(grr))
resultAnnot <- grr
all(rownames(dds_rnaseq) == rownames(resultAnnot))
#make the nice names
nice_names <- paste(resultAnnot$ensembl_gene_id,
                    resultAnnot$external_gene_name, 
                    sep = '_')
resultAnnot$nice_names <- nice_names
head(resultAnnot)
all(rownames(dds_rnaseq) == rownames(resultAnnot))
#check names
rld_rnaseq <- rlog(dds_rnaseq, blind = TRUE)
idx2 <- match(rownames(result_sex_selected)[1:50],rownames(dds_rnaseq))
plotme <- (rld_rnaseq)[rownames(result_sex_selected)[1:50],]
rownames(plotme) <- resultAnnot$nice_names[idx2]

#make heatmap with candidate genes
pheatmap(assay(plotme),
         scale="row",
         fontsize_row=10,
         cellheight=12, cellwidth=12,
         main="Heatmap of candidate genes")

```


Marker gene Profiles

```{r marker_gene_profiles, echo=FALSE}
#But what about the marker profiles from known markers?
#Load some developmental markers to assess the expression profile
mkr <- read.table("markers.txt",
                 stringsAsFactors = F,
                 fill=T, row.names = 1,
                 header=T, sep="\t")
rownames(mkr)
#%in%  This makes sure the marker gene names are rownames in the object
valid_names <- rownames(mkr)[rownames(mkr) %in% rownames(dds_rnaseq)]
plotmetoo <- rld_rnaseq[valid_names,]
idx3 <- match(rownames(plotmetoo),rownames(dds_rnaseq))
rownames(plotmetoo) <- resultAnnot$nice_names[idx3]
pheatmap(assay(plotmetoo),
         scale="row",
         fontsize_row=10,
         cellheight=12, cellwidth=12,
         treeheight_row=40, treeheight_col=40,
         main="Heatmap of developmental marker genes")

```

Finally add column annotation

```{r column_annotation, echo=FALSE}
#put in column annotation
mkr_bit <- mkr[valid_names,]
rownames(mkr_bit) <- resultAnnot$nice_names[idx3]

Sys.setlocale("LC_ALL", "C")
pheatmap(assay(plotmetoo),
         scale="row",
         fontsize_row=10,
         annotation_row=as.data.frame(mkr_bit[,c("LongState","ShortState")]),
         cellheight=12, cellwidth=12,
         treeheight_row=40, treeheight_col=40,
         main="Heatmap of developmental marker genes (with col annotation)")

```

Additional Reading
http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html

## Session Information
```{r session_info, echo=FALSE}

sessionInfo()

```


